crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -46,10 +46,23 @@ impl<'db> UnionBuilder<'db> {
     pub(crate) fn add(mut self, ty: Type<'db>) -> Self {
         match ty {
             Type::Union(union) => {
-                self.elements.extend(union.elements(self.db));
+                for element in union.elements(self.db) {
+                    self = self.add(*element);
+                }
             }
             Type::Never => {}
             _ => {
+                let mut remove = vec![];
+                for element in &self.elements {
+                    if ty.is_subtype_of(self.db, *element) {
+                        return self;
+                    } else if element.is_subtype_of(self.db, ty) {
+                        remove.push(*element);
+                    }
+                }
+                for element in remove {
+                    self.elements.remove(&element);
+                }
```

Comment:
Does it still make sense to use a hash set for `elements`, now that the "uniqueness" is no longer determined by `Eq` and `Hash` but by `is_subtype_of`? I get the impression that we could as well use a `Vec`.

Reply:
This explains why mypy uses a Python `list` for the inner list of types in its `Union` representation, rather than a Python `set` (something I've been wondering about recently)